Support --storage-opt for aufs graph driver

dqminh opened the pull request:
This patch alllows user to pass extra mount options to aufs graph driver via the daemon's `--storage-opt` flag. 

Our use case is that we want to test `dirperm1` mount option ( http://aufs.sourceforge.net/aufs3/man.html ) which only available to kernel with newer aufs patches to see if it can mitigate the permission bug with aufs (https://github.com/docker/docker/issues/783)

dqminh committed:
add support for aufs.mountopt

Docker-DCO-1.1-Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com> (github: dqminh)

dqminh committed:
add a README to document aufs.mountopt

Docker-DCO-1.1-Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com> (github: dqminh)

dqminh committed:
nit: correct function name of Init comments

Docker-DCO-1.1-Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com> (github: dqminh)

dqminh committed:
update docker man page to document aufs storage options

Docker-DCO-1.1-Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com> (github: dqminh)

dqminh committed:
add documentation for aufs.mountopt in cli

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

jfrazelle commented:
can we add unit tests for this

dqminh commented:
@jfrazelle hmm i'm not sure. Is there a way to run a specific test only when the underlying storage is aufs now ?

jfrazelle commented:
Couldn't we add something to aufs_test.go the unit test, to at least parse
the flag

On Wednesday, January 7, 2015, Daniel, Dao Quang Minh <
notifications@github.com> wrote:

> @jfrazelle <https://github.com/jfrazelle> hmm i'm not sure. Is there a
> way to run a specific test only when the underlying storage is aufs now ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9684#issuecomment-68995631>.
>

dqminh committed:
add test for initializing aufs.mountopt

Docker-DCO-1.1-Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com> (github: dqminh)

SvenDowideit commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -329,6 +334,21 @@ Currently supported options are:
+
+    Example use:
+
+        # tell aufs to checks the permission bits of the directory on the
```
`tell aufs to check` (no `s`)

SvenDowideit commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -329,6 +334,21 @@ Currently supported options are:
+        # tell aufs to checks the permission bits of the directory on the
+        # topmost branch and ignore the permission bits on all lower branches
+        # this option is only available for kernel with dirperm1 patches
+        $ sudo docker -d --storage-opt aufs.mountopt=dirperm1
```
if you can explain why a user would want to do that, that would be awesome :)

SvenDowideit commented:
small nits, LGTM when resolved - @fredlf @jamtur01

dqminh commented:
@SvenDowideit i added some more information about `dirperm1` mount option in `cli.md`. PTAL.

SvenDowideit commented on daemon/graphdriver/aufs/README.md:
```diff
@@ -0,0 +1,25 @@
+
+The aufs graph driver uses AUFS (advanced multi layered unification filesystem)
+to implement COW layers. For each graph location ( typically stored at
+/var/lib/docker/aufs ), three directories are created:
```
can you put the directory into ``` code ticks please?

SvenDowideit commented on daemon/graphdriver/aufs/README.md:
```diff
@@ -0,0 +1,25 @@
+to implement COW layers. For each graph location ( typically stored at
+/var/lib/docker/aufs ), three directories are created:
+
+- layers: contains the metadata files for each layers. Each metatada file
```
same - use `layers` as a dir name.

SvenDowideit replied:
`metatada` :)

SvenDowideit commented on daemon/graphdriver/aufs/README.md:
```diff
@@ -0,0 +1,25 @@
+/var/lib/docker/aufs ), three directories are created:
+
+- layers: contains the metadata files for each layers. Each metatada file
+  contains its layer's parent ids.
```
I think we capitalise `ID`

SvenDowideit commented on daemon/graphdriver/aufs/README.md:
```diff
@@ -0,0 +1,25 @@
+- diff: contains content of the layers. A layer's content will be saved into a
+  directory with the layer's id as the directory name.
+
+- mnt: mount points of the rw layers.
```
mmm, please expand `rw` to read-write

SvenDowideit commented on daemon/graphdriver/aufs/README.md:
```diff
@@ -0,0 +1,25 @@
+
+- mnt: mount points of the rw layers.
+
+### options
```
Capitalise the heading :)

SvenDowideit commented on daemon/graphdriver/aufs/README.md:
```diff
@@ -0,0 +1,25 @@
+
+### options
+
+The aufs backend supports some options that you can specify when starting the
```
i think aufs should be in backticks too

SvenDowideit commented on docs/man/docker.1.md:
```diff
@@ -292,6 +297,12 @@ for data and metadata:
       --storage-opt dm.metadatadev=/dev/vdc \
       --storage-opt dm.basesize=20G
 
+Launching docker daemon with *aufs*backend that applies `dirperm1` mount option
```
missing a space after `*aufs*`

fredlf replied:
Capitalize Docker

SvenDowideit commented:
nice :) I'm still LGTM, though now with fixes :)

fredlf commented on daemon/graphdriver/aufs/README.md:
```diff
@@ -0,0 +1,25 @@
+## aufs - a storage backend based on AUFS
+
+The aufs graph driver uses AUFS (advanced multi layered unification filesystem)
```
multi-layered

fredlf commented on daemon/graphdriver/aufs/README.md:
```diff
@@ -0,0 +1,25 @@
+## aufs - a storage backend based on AUFS
+
+The aufs graph driver uses AUFS (advanced multi layered unification filesystem)
+to implement COW layers. For each graph location ( typically stored at
```
delete extra space after (

fredlf commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -179,14 +179,17 @@ The `aufs` driver is the oldest, but is based on a Linux kernel patch-set that
 containers to share executable and shared library memory, so is a useful choice
-when running thousands of containers with the same program or libraries.
+when running thousands of containers with the same program or libraries. AUFS
+driver can be configured with `--storage-opt` flags. Refer to [AUFS
```
The AUFS driver...

fredlf commented on docs/man/docker.1.md:
```diff
@@ -219,8 +219,8 @@ inside it)
-only backend which currently takes options is *devicemapper*. Therefore use these
-flags with **-s=**devicemapper.
+backends which currently take options are *devicemapper* and *aufs*. Therefore
+use these flags with **-s=**devicemapper or **-s=**aufs
```
Comma after "therefore".

fredlf commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -331,6 +336,24 @@ Currently supported options are:
+        # topmost branch and ignore the permission bits on all lower branches.
+        # This option is only available for kernel with `dirperm1` patches and
+        # can be used to fix aufs' permission bug (i.e. upper layer having
+        # broader mask than the lower layer). More information about the
```
Comma after i.e., please.

fredlf commented:
Docs LGTM once my copy-edits are fixed.

fredlf commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -331,6 +336,24 @@ Currently supported options are:
 
+##### AUFS storage driver options
+
+The aufs options are prefixed with `aufs`. Current supported options are:
```
Currently

fredlf commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -331,6 +336,24 @@ Currently supported options are:
+
+ * `aufs.mountopt`
+
+    Specifies extra mount options used when mounting the aufs layers.
```
Incomplete sentence

crosbymichael commented:
LGTM

@tiborvass is going to fix the doc changes and get this one merged

tiborvass commented:
Carried in #10390
